### ISSUE TITLE
[Explore Source Code] Run sklearn clustering + lof for each cluster

### ISSUE THREAD

**Issue by victorcheong** [victorcheong/da_toolkit#2]
**Action Item**
Adopt a 2 step process to identify outliers and compare runtime & accuracy with a single run of sklearn LOF algorithm.
- Step 1: Use K-means clustering to cluster the data points.
- Step 2: For each cluster, perform sklearn LOF.

**Comment by victorcheong** [comment]
For all experiments below, the chosen clustering algorithm is mini-batch kmeans. The number of clusters is 50.
# 60 million rows
The percentage of anomalies is 0.1%. The contamination parameter for the LOF for each cluster will be set as 0.1%.
## Global Outliers
| Metric | 2 step process (Clustering + Sklearn LOF) | Sklearn LOF |
| :---:   | :---: | :---: |
| Runtime | Clustering: 32.7 seconds <br /> LOF: 349.2 seconds <br /> Total: **381.9 seconds**| **317.5 seconds**   |
| F1 Score | 74.7%  | 74.7%   |
## Local Outliers
| Metric | 2 step process (Clustering + Sklearn LOF) | Sklearn LOF |
| :---:   | :---: | :---: |
| Runtime | Clustering: 83.7 seconds <br /> LOF: 246.6 seconds <br /> Total: **330.3 seconds**| **291.4 seconds**   |
| F1 Score | 95.7%  | 95.6%   |

For the next few experiments on the real world datasets, the contamination parameter will be based on the percentage of anomalies in the training/test sets (does not matter since it is a stratified split).
# ML Matt Dataset
| Metric | 2 step process (Clustering + Sklearn LOF) | Sklearn LOF |
| :---:   | :---: | :---: |
| Runtime | Clustering: 0.2 second <br /> LOF: 0.1 second <br /> Total: **0.3 second**| **2.8 seconds**   |
| F1 Score | 28.3%  | 28.5%   |
# Access Behaviour Anomaly Dataset
| Metric | 2 step process (Clustering + Sklearn LOF) | Sklearn LOF |
| :---:   | :---: | :---: |
| Runtime | Clustering: 0.2 second <br /> LOF: < 0.1 second <br /> Total: **< 0.3 second**| **< 0.1 second**   |
| F1 Score | 32.8%  | 28.2%   |
# Energy
| Metric | 2 step process (Clustering + Sklearn LOF) | Sklearn LOF |
| :---:   | :---: | :---: |
| Runtime | Clustering: 0.2 second <br /> LOF: 0.1 second <br /> Total: **0.3 second**| **0.6 second**   |
| F1 Score | 32.3%  | 34.3%   |

From the experiments done above, it is inconclusive to determine if the runtime for the 2 step process is slower than a single pass of LOF. F1 scores are largely similar. A single pass of LOF is preferred as there are less codes to run.

To validate if the above is true, the algorithms are compared on larger synthetic datasets.

# Global Outliers
| Size of dataset | 2 step process (Clustering + Sklearn LOF) | Sklearn LOF |
| :---:   | :---: | :---: |
| 70 million rows | Runtime: 95.9 seconds + 376.9 seconds = **472.8 seconds** | Runtime: **353.9 seconds** |
| 80 million rows | Runtime: 89.3 seconds + 484.7 seconds = **574 seconds** | Runtime: vs code crashes |

Further testing was done to see the largest dataset size that the 2 step process can handle. 
| Size of dataset | 2 step process (Clustering + Sklearn LOF) | Sklearn LOF |
| :---:   | :---: | :---: |
| 70 million rows | Runtime: 95.9 seconds + 376.9 seconds = **472.8 seconds** | Runtime: **353.9 seconds** |
| 80 million rows | Runtime: 89.3 seconds + 484.7 seconds = **574 seconds** | Runtime: vs code crashes |
| 100 million rows | Runtime: 160.3 seconds + 594.8 seconds = **755.1 seconds** | Runtime: vs code crashes |
| 150 million rows | Runtime: 178.1 seconds + 888.7 seconds = **1066.8 seconds** | Runtime: vs code crashes |
| 200 million rows | Runtime: 108.9 seconds + 1218.4 seconds = **1327.3 seconds** | Runtime: vs code crashes |

The dataset for 250 million rows took too long to be created. As such, the experiment is stopped at 200 million rows. 

**The conclusion is that for smaller datasets, the 2 step process will take longer to run than a single pass LOF, but at least it can handle larger datasets.**

# References
Refer to [document](https://github.com/victorcheong/da_toolkit/blob/master/documentation/explore_source_code/Results%20on%20open-source%20datasets.pdf) -> Experiment 13 (Sklearn clustering + LOF)

Codes
Refer to [code](https://github.com/victorcheong/da_toolkit/blob/master/documentation/explore_source_code/code.ipynb)-> Experiment 13 (Sklearn clustering + LOF)
